aiosmtpd/smtp.py

```diff
@@ -591,6 +593,32 @@ def _syntax_available(self, method):
         status = await self._call_handler_hook('RSET')
         await self.push('250 OK' if status is MISSING else status)
 
+    def _parse_xclient_attributes(self, attributes):
+        """
+        Parse key-value arguments and return OrderedDict on success.
+        Raise ValueError on wrong arguments.
+        """
+        result = collections.OrderedDict()
+        for kv in attributes.split(' '):
+            key, value = kv.split('=')
+            if key not in XCLIENT_ATTRIBUTES:
+                raise ValueError()
+            result[key] = value
+        return result
+
+    @syntax('XCLIENT [key1=value1] ... [keyN=valueN]')
+    async def smtp_XCLIENT(self, arg):
+        if not hasattr(self.event_handler, 'handle_XCLIENT'):
```

Comment:
I think I'm okay with this, and I understand why you're doing it this way.  Of course, it needs to be documented (and in fact, there should be a NEWS entry and documentation to go along with this change).

It's a *little* weird though, because the server will claim to support `XCLIENT` in cases where it actually doesn't because the handler doesn't support it.  Ideally, `XCLIENT` wouldn't show up in the ESMTP response unless it was actually supported.  What do you think about that?

The alternative is to do everything except include the `smtp_XCLIENT()` method, and then document how a subclass would implement and enable it.

Reply:
I agree that this is weird. I'll try to return XCLIENT in ESMTP response only if corresponding handler method exists.